GitHubRepositories: Add Output Type Views

X-Guardian opened the pull request:
#### Pull Request description

This PR adds output type views for the following resources in the `GitHubRepositories` module:

- `GitHub.Repository`
- `GitHub.RepositoryTopic`
- `GitHub.RepositoryContributor`
- `GitHub.RepositoryContributorStatistics`
- `GitHub.RepositoryCollaborator`
- `GitHub.RepositoryTag`

#### Example Views

##### GitHub.Repository

```powershell
Get-GitHubRepository -OwnerName 'octocat' | Select-Object -First 5

full_name                  visibility description
---------                  ---------- -----------
octocat/boysenberry-repo-1 public     Testing
octocat/git-consortium     public     This repo is for demonstration purposes only.
octocat/hello-worId        public     My first repository on GitHub.
octocat/Hello-World        public     My first repository on GitHub!
octocat/linguist           public     Language Savant. If your repository's language is being reported incorrectly, send us a pull request!
```

##### GitHub.RepositoryTopic

```powershell
Get-GitHubRepositorytopic -OwnerName 'microsoft' -RepositoryName PowerToys

names : {powertoys, desktop, windows}
RepositoryUrl: https://github.com/microsoft/PowerToys
```

##### GitHub.RepositoryContributor

```powershell
Get-GitHubRepositorycontributor -OwnerName 'Microsoft' -RepositoryName PowerShellForGithub | Select-Object -First 5

UserName       type contributions
--------       ---- -------------
KarolKaczmarek User 122
HowardWolosky  User 77
joseartrivera  User 7
jpomfret       User 5
jrolstad       User 3
```

##### GitHub.RepositoryContributorStatistics

```powershell
Get-GitHubRepositorycontributor -OwnerName 'Microsoft' -RepositoryName PowerShellForGithub -IncludeStatistics | Select-Object -First 5

author.UserName total weeks
--------------- ----- -----
wikijm          1     {@{w=1457827200; a=0; d=0; c=0}, @{w=1458432000; a=0; d=0; c=0}, @{w=1459036800; a=0; d=0; c=0}, @{w=1459641600; a=0; d=0; c=0}…}
v2kiran         1     {@{w=1457827200; a=0; d=0; c=0}, @{w=1458432000; a=0; d=0; c=0}, @{w=1459036800; a=0; d=0; c=0}, @{w=1459641600; a=0; d=0; c=0}…}
rjmholt         1     {@{w=1457827200; a=0; d=0; c=0}, @{w=1458432000; a=0; d=0; c=0}, @{w=1459036800; a=0; d=0; c=0}, @{w=1459641600; a=0; d=0; c=0}…}
mtboren         1     {@{w=1457827200; a=0; d=0; c=0}, @{w=1458432000; a=0; d=0; c=0}, @{w=1459036800; a=0; d=0; c=0}, @{w=1459641600; a=0; d=0; c=0}…}
kant            1     {@{w=1457827200; a=0; d=0; c=0}, @{w=1458432000; a=0; d=0; c=0}, @{w=1459036800; a=0; d=0; c=0}, @{w=1459641600; a=0; d=0; c=0}…}
```

##### GitHub.RepositoryCollaborator

```powershell
Get-GitHubRepositoryCollaborator -OwnerName 'dsccommunity' -RepositoryName ActiveDirectoryDsc | Select-Object -First 5

UserName        permissions.admin permissions.push permissions.pull
--------        ----------------- ---------------- ----------------
tysonjhayes     False             False            True
NikCharlebois   False             False            True
concentrateddon False             False            True
mgreenegit      True              True             True
johlju          True              True             True
```

##### GitHub.RepositoryTag

```powershell
Get-GitHubRepositoryTag -OwnerName 'microsoft' -RepositoryName PowerShellForGithub | Select-Object -First 5

name   commit.sha
----   ----------
0.14.0 ed9b114fe25dfb2d91752cbf06d0b5c0389b30ca
0.13.1 d279f1c7e1bd8261cdcf9dab6a8b78d3a0004457
0.13.0 79c30d4d3ce2b56281ab6c040173973c416b2ef2
0.12.0 d68503ffc0c3e16f515871570e80aecf2ac9693e
0.11.0 78cd480e1ee9f5baa9ce1d58462210bf8bfe9087
```

HowardWolosky commented on GitHubRepositories.ps1:
```diff
@@ -1340,7 +1442,19 @@ function Get-GitHubRepositoryLanguage
     }
 
-    return Invoke-GHRestMethodMultipleResult @params
+    $multipleResult = Invoke-GHRestMethodMultipleResult @params
+    foreach ($result in $multipleResult)
+    {
+        $result.PSObject.TypeNames.Insert(0,$script:GitHubRepositoryLanguageTypeName)
+    }
+
+    $multipleResult = Invoke-GHRestMethodMultipleResult @params
+    foreach ($result in $multipleResult)
+    {
+        $result.PSObject.TypeNames.Insert(0,$script:GitHubRepositoryLanguageTypeName)
+    }
+
+    return $multipleResult
```
Looks like you have this duplicated.

```suggestion
    $multipleResult = Invoke-GHRestMethodMultipleResult @params
    foreach ($result in $multipleResult)
    {
        $result.PSObject.TypeNames.Insert(0, $script:GitHubRepositoryLanguageTypeName)
    }

    return $multipleResult
```

X-Guardian replied:
Whoops, Done.

HowardWolosky commented on GitHubRepositories.ps1:
```diff
@@ -1,6 +1,13 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
+$script:GitHubRepositoryTypeName = 'GitHub.Repository'
+$script:GitHubRepositoryTopicTypeName = 'GitHub.RepositoryTopic'
+$script:GitHubRepositoryContributorTypeName = 'GitHub.RepositoryContributor'
+$script:GitHubRepositoryCollaboratorTypeName = 'GitHub.RepositoryCollaborator'
+$script:GitHubRepositoryLanguageTypeName = 'GitHub.RepositoryLanguage'
+$script:GitHubRepositoryTagTypeName = 'GitHub.RepositoryTag'
```
Might be worthwhile to initialize these as read-only script vars the way we do in GitHubCore:

```suggestion
@{
    GitHubRepositoryTypeName = 'GitHub.Repository'
    GitHubRepositoryTopicTypeName = 'GitHub.RepositoryTopic'
    GitHubRepositoryContributorTypeName = 'GitHub.RepositoryContributor'
    GitHubRepositoryCollaboratorTypeName = 'GitHub.RepositoryCollaborator'
    GitHubRepositoryLanguageTypeName = 'GitHub.RepositoryLanguage'
    GitHubRepositoryTagTypeName = 'GitHub.RepositoryTag'
 }.GetEnumerator() | ForEach-Object {
     Set-Variable -Scope Script -Option ReadOnly -Name $_.Key -Value $_.Value
 }
```

X-Guardian replied:
Done.

As an aside, this repo has `alignPropertyValuePairs` as `false` in its vscode settings. Personally, I think they are much easier to read when aligned i.e.:

```powershell
@{
    GitHubRepositoryTypeName             = 'GitHub.Repository'
    GitHubRepositoryTopicTypeName        = 'GitHub.RepositoryTopic'
    GitHubRepositoryContributorTypeName  = 'GitHub.RepositoryContributor'
    GitHubRepositoryCollaboratorTypeName = 'GitHub.RepositoryCollaborator'
    GitHubRepositoryLanguageTypeName     = 'GitHub.RepositoryLanguage'
    GitHubRepositoryTagTypeName          = 'GitHub.RepositoryTag'
 
}.GetEnumerator() | ForEach-Object {
    Set-Variable -Scope Script -Option ReadOnly -Name $_.Key -Value $_.Value
}
```
```

HowardWolosky replied:
I've been on both sides of that argument over the years at Microsoft on different teams.  I'm currently on the side of _not_ doing that.  I _totally_ agree that in isolation, that's easier to read.  It mostly comes down to maintenance.  If it's likely that values are to be added/removed over time, then it becomes annoying both from a maintainer and code reviewer when having to deal with re-adjusting the spacing within there as things are added and removed.  And even if _this_ one is likely to remain more-or-less static, it then adds precedent/pressure to do so in other parts of the code that may be less static, or the code then starts to lack conformity.

X-Guardian replied:
That's the beauty of modern editors like VSCode, Shift-Alt-F and it formats it for you!

HowardWolosky commented on PowerShellForGitHub.Format.ps1xml:
```diff
@@ -0,0 +1,145 @@
+                <PropertyName>full_name</PropertyName>
+              </TableColumnItem>
+              <TableColumnItem>
+                <PropertyName>private</PropertyName>
```
```suggestion
                <PropertyName>visibility</PropertyName>
```

It might be more readable to have this column just be "visibility" and then have a scriptblock for the TableColumnItem that prints 'public' or 'private' based on the state.

X-Guardian replied:
I've just changed the column to `visibility`. Don't think there is any need for a script block?

HowardWolosky replied:
I think I wrote that comment either before I actually reviewed your other PR that introduced the `Visibility` property, or I simply forgot about it.  Correct...given that there is an actual `Visibility` property that uses the words we want, there's no need for a scriptblock to accomplish that task.

HowardWolosky commented on GitHubRepositories.ps1:
```diff
@@ -175,7 +189,10 @@ function New-GitHubRepository
 
-    return Invoke-GHRestMethod @params
+    $result =  Invoke-GHRestMethod @params
+    $result.PSObject.TypeNames.Insert(0,$script:GitHubRepositoryTypeName)
```
nit: Here and elsewhere, can you add a space after the comma

```suggestion
    $result.PSObject.TypeNames.Insert(0, $script:GitHubRepositoryTypeName)
```

X-Guardian replied:
Done

X-Guardian commented:
I've moved the `ps1xml` format file into a `Formatters` directory, and renamed it, so that there can be one format file for each nested module.

HowardWolosky commented:
Thanks -- will give a final pass at this tomorrow, and then hopefully get it merged in.

HowardWolosky commented:
I'm holding off on this one for another couple days as I'm examining a consistent approach to pipelining and type definitions across the entire module.  We'll _definitely_ be taking this formatter change soon.  It's just that the work you're doing to add the type information in this PR may be superseded by that larger effort (which is definitely influenced by what you're doing here).

If interested, you can take a look at my early thoughts on that larger approach: HowardWolosky/PowerShellForGitHub@e6b5ef695e99aebf9a02bed557019d3e3f5a3800

X-Guardian commented:
No problem @HowardWolosky.

X-Guardian commented:
I'll finish this PR once all the other `GitHubRepositories` PRs have been merged: #221, #233 and #235.

X-Guardian commented:
@HowardWolosky, just looking at finishing this PR. The current output types of `Get-GitHubRepositoryCollaborator` and `Get-GitHubRepositoryContributor` have been set as `GitHub.User`, but they return a different schema than `Get-GitHubUser`, so they need to have their own types. 

- [Get Repository Collaborator Response](https://developer.github.com/v3/repos/collaborators/#response)
- [Get Repository Contributor Response](https://developer.github.com/v3/repos/#response-if-repository-contains-content)
- [Get GitHub User Response](https://developer.github.com/v3/users/#response)

Are you OK with `GitHub.RepositoryCollaborator` and `GitHub.RepositoryContributor`?

HowardWolosky commented:
Nice catch.  That was an oversight on my part.  I didn't look carefully enough.

> Are you OK with `GitHub.RepositoryCollaborator` and `GitHub.RepositoryContributor`?

I'm feeling a bit torn. Those make complete sense to me on the one hand. On the other, `Repository` feels a little redundant because that's the only thing someone could be a contributor or collaborator on.  That then puts into question whether `Get-GitHubRepository[Contributor|Collaborator]` should really be `Get-GitHub[Contributor|Collaborator]`.

I'd say move forward with your suggestion.  If it feels unnatural after we start using it, we can look at changing it.

Thanks again.  I _really_ appreciate all of your contributions to this project.

X-Guardian commented:
OK, code updated. I've changed the `login` property to the `UserName` property on the `GitHub.RepositoryContributor` and `GitHub.RepositoryCollaborator` type views and added the `RepositoryUrl` to the `GitHub.RepositoryTopic` type view.

X-Guardian commented:
@HowardWolosky, can we trigger the CI for this PR?

HowardWolosky commented on GitHubRepositories.ps1:
```diff
@@ -2850,3 +2853,189 @@ filter Add-GitHubRepositoryAdditionalProperties
+            $UserName = $item.login
+            if ([String]::IsNullOrEmpty($UserName) -and $PSBoundParameters.ContainsKey('Name'))
+            {
+                $UserName = $Name
```
nit: camelCase since it's a local var.  Here and with its reference below.  Same for `UserId`.  Same for other function above.

```suggestion
                $userName = $Name
```

X-Guardian replied:
Done

HowardWolosky commented:
/azp run PowerShellForGitHub-CI
